scanner/scanner.go

```diff
@@ -0,0 +1,168 @@
+package scanner
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+)
+
+// constants
+const (
+	EOF = iota
+	Whitespace
+	LeftParen
+	RightParen
+	LeftBracket
+	RightBracket
+	Semicolon
+	Backslash
+	String
+	Illegal
+)
+
+var eof = rune(0)
+
+// isWhiteSpace is a check to see if a rune is whitespace
+// (space, tab, newline, carriage return)
+func isWhitespace(ch rune) bool {
+	return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r'
+}
+
+// isSpecial is a check for special characters used in SGF files
+func isSpecial(ch rune) bool {
+	return ch == '(' || ch == ')' || ch == '[' || ch == ']' || ch == ';'
+}
+
+// Token contains attributes for the Type of token and the raw string
+type Token struct {
+	Type int
+	Raw  string
+}
+
+// Scanner wraps bufio.Reader but has the public Scan() function to scan tokens
+type Scanner struct {
+	*bufio.Reader
+}
+
+// New creates a scanner object out of an io.Reader interface
+func New(r io.Reader) *Scanner {
+	return &Scanner{bufio.NewReader(r)}
+}
+
+// Either return the current rune or EOF (null)
+func (s *Scanner) read() rune {
+	ch, _, err := s.ReadRune()
+	if err != nil {
+		return eof
+	}
+	return ch
+}
+
+// Go back one rune
+func (s *Scanner) unread() {
+	s.UnreadRune()
+}
+
+// Scan is the function that tokenizes the input and outputs tokens
+func (s *Scanner) Scan() *Token {
```

Comment:
nit: I'd recommend passing back the number of chars read so the caller can can keep track of of the column / line # for error messaging. Or maybe # of chars and # of newlines.